### ISSUE TITLE
Evaluate Local State Implementation

### ISSUE THREAD

**Issue by a-laughlin** [FormidableLabs/urql#462]
Nice work with urql's extensibility and small core!

I'm interested in evaluating and potentially maintaining a local state implementation.  The only discussion looks like #323.  Any other thought on it so far?

**Comment by kitten** [comment]
The only reason why we haven’t gone for this is that we think there’s probably a limit to how many exchanges we can maintain ourselves 😅

So if you’re willing to give it a shot, please go for it! I’m always willing to help out and support you (Twitter DMs are open)

There’s probably a lot of different ways to approach this too. You could either go for adding directives that are handled by an exchange and stripped out when it passes on operations, or you can go for a centralised configuration and then selectively strip out stuff that is handled by local resolvers.

**Comment by a-laughlin** [comment]
Fantastic.  Thanks for the insights Phil.

My inclination is to start simple with a convenience wrapper around the exchange and result creation.  Here's some pseudocode for an exchange and usage.  See any reasons the concept wouldn't work for an initial effort?

```js
// factory
const makeLocalResolverExchange = resolverFn =>{
  return ({ forward }) => {
    return operation$ => {
      const resolver = resolverFn( ()=>forward(operation$) );
      return pipe(
         operations$,
         tryCatch(
           map( op=>makeOperationResult( resolver(op) ) )
           makeErrorResult
         )
       );
      }
    }
  };
};

// usage
const localResolverExchange = makeLocalResolverExchange(
  forwardOperation=>operation=>{
    return cond([
      [userPredicateA,syncResolverA],
      [userPredicateB,syncResolverB],
      [userPredicateC,asyncResolverC],
      [stubTrue,forwardOperation]
    ])(operation)
  )
);
```

notes:
- assumes exchanges: [dedupe,cache,localResolvers,fetch]
- can return synchronous or asynchronous results
- ignores requestPolicy since caches handle it
- Tradoffs: At the cost of making the user responsible to parse the query/mutation/subscription request and return a valid graphql response shape, user gets flexibility, no need for custom directives, and easy to add/remove some/all resolvers for mocking and local state.  Maintainers get little to maintain.
- Would need docs on query/mutation/subscription parsing.  Happy to write those.  Some helper functions might be handy.  TBD on that.
- Docs should also mention that separate cache and localResolver exchanges means a "cache-only" requestPolicy will not call local resolvers.  An all-local-data setup would involve a "cache-first" policy plus local resolvers that handle all requests.

**Comment by a-laughlin** [comment]
Reflecting, it might help to clarify the problem a bit.  "Local State" is too general a name given that all exchanges handle data state in some way.  The problem this issue addresses is more how to implement a local version of a server.  Some initial requirements for that:
- provides data in user-specified conditions
- allows forward to exchanges that handle other requests (eg fetching).
- allows custom operations
- lightweight since it's on the client, so minimal dependencies
- reads and writes user-specified data
- returns expected data shape
- returns server-like errors
- doesn't handle caching by itself
- supports queries, mutations, subscriptions

On a related note, the graphcache exchange appears to have resolvers.  I'm curious about their intent, especially since the default cache lacks them.  Resolvers on a cache seem likely to mix the concerns of caching with creating/updating/removing/transforming/deriving.  The way I see this exchange would be handling those concerns while providing responses that both the default cache and graph cache could digest effectively.

**Comment by JoviDeCroock** [comment]
In theory, the graphcache _could_ handle local state if it was aware of `@local` directives and recognized them but yes, wasn't our original intent there.

**Comment by a-laughlin** [comment]
Thanks for the clarification Jovi.  Reading through all the graphcache docs now.

**Comment by a-laughlin** [comment]
Quick update.  I'm assembling a spreadsheet with all the exchange responsibilities by requestPolicy and operation type to get a sense of what cases the local (state|data|something) exchange should handle.  https://docs.google.com/spreadsheets/d/14tsuPWvVvoG-zKzx9zvDbBFEP61X2t4mevRTZJRplwQ/edit#gid=0

Getting through it is going to take some time (though it's a great way to get familiar with the ecosystem).

**Comment by JoviDeCroock** [comment]
You can probably make this work for both caches in the following way:
- you put an exchange in front of cache
- it analyses incoming query/mutation
- takes the local directives
- executes by getting data and so on
- forwards to cache without local refs
- when operation is returning from the end attach relevant data

**Comment by a-laughlin** [comment]
Thanks for thinking through that approach Jovi.  I don't have time this evening to chew on it and respond, but should sometime in the next few days.

My own thoughts aside, one question has me curious.  As Phil mentioned earlier in the thread "You could either go for adding directives (me: like `@local`) that are handled by an exchange and stripped out when it passes on operations, or you can go for a centralised configuration and then selectively strip out stuff that is handled by local resolvers".  If I understand the graph cache correctly, it already does the second option.  In addition to per per-object caching, it does transformations, updates, selections of subsets, and merging partial responses.  It seems like it's only missing an "initial state" property, and enabling to resolvers to intercept mutations, to function as local state manager.  Why not go that route?

**Comment by a-laughlin** [comment]
Re `@local` directives, I'm opposed to architectures using them for a few reasons. They:
- potentially impact other exchanges
- require additional code to strip out
- create a sequential exchange order dependency that can break unexpectedly
- require additional error handling for cases like that
- require resolvers to work
- require a change in two places to handle one query (resolver + query)
- fail silently if the resolver exchange is missing (assuming it doesn't make the other exchanges produce errors)
- require users to learn another concept
- require users to learn  additional syntax
- create two places to check when debugging
- require resolver documentation
- require directive documentation
- decrease query/mutation reusability

To contrast, an exchange of resolvers that intercept requests:
- require resolvers to work
- require resolver documentation

I don't believe directives' benefits outweigh their cons.  However, I'm open to changing my mind on that.

**Comment by a-laughlin** [comment]
Here's where my thinking is at currently:

Terminology: Fetch, dedupe, debug - these are verbs.  "Local State" is adjective+noun.  The distinction is important because nouns don't "do" anything.  It's easy to lump a bunch of functionality into code based on them, so I wanted a verb.

The essence of local state is mapping requests/responses to/from a state object(s), and transitioning (often reducing) that object from one state to another.  As such, I created two tiny composable exchanges, map and reduce.  Users can compose them in multiple pipeline locations depending on their needs.  For example:

```js
client({
  exchanges: [
    mapExchange({ ...}), // stateless sync non-cached: arbitrary transforms/selections/filters/derivations
    reduceExchange({ ...}), // stateful sync non-cached: arbitrary transforms/selections/filters/derivations
    cacheExchange,
    mapExchange({ ...}), // stateless sync cached: arbitrary transforms/selections/filters/derivations
    reduceExchange({ ...}), // stateful sync cached: arbitrary transforms/selections/filters/derivations
    graphCacheExchange, // intelligently cache gql-valid data
    mapExchange({ ...}), // stateless async cached: gql-valid transforms
    reduceExchange({ ...}), // stateful async cached: gql-valid transforms
    fetchExchange,
  ]
})
```

In that example, "local state" occurs in three reduce and two cache exchanges, each with its own purpose.  That approach composes flexibly with urql's other exchanges, and provides users with a couple of small exchanges they can extend to build more customized solutions (e.g., the reduce exchange is built using the map exchange).

Initial concept code for [map](https://github.com/a-laughlin/urql/blob/master/src/exchanges/map.ts) and [reduce](https://github.com/a-laughlin/urql/blob/master/src/exchanges/reduce.ts) is up on my fork for review.  I pushed  (pushed with --no-verify due to a pre-commit hook syntax issue I couldn't quickly resolve).  That issue, my infrequent typescript use, and this being my first exchange, add up to the code likely having some problems.

Once we get to an architecture we all like, I'll add tests and docs.  "Local State" would be one example use case in the reduce docs.

Thoughts on that approach?

**Comment by kitten** [comment]
I’m opposed to directives as well, when there’s good alternatives, since they need the definition of data requirements, ie the query, to know what’s local data.

However, I don’t think the assessment here is quite right for what local state implies, sorry 😅

> require additional code to strip out [fields that are for local state]

This is required either way, as we need to strip parts of a query that are for a local schema or state and not the API / remote schema

I have taken a look at your `map` and `reduce` exchanges, but I think they’re unfortunately too small to be published as part of the core. I personally don’t see exchanges as a DSL so introducing another abstraction on top of Wonka streams (which are basically Rx-inspired) would add much value as these exchanges are very small.

Basically stateless exchanges are just utilities that don’t have to live in core as they’re very reproducible and just a group of a couple of Wonka operators on top of our convention. IMHO that would add more of a burden to users as there’d be more docs and more concepts that look like conventions they have to adhere to.

A local state exchange is also inherently stateful as far as I can tell 🤔

I think the way to go here is a full resolver/schema exchange for local state that allows CRUD getters/setters to be defined and strips out fields that it handles before passing these stripped queries on to the `fetchExchange` / `cacheExchange`

**Comment by a-laughlin** [comment]
> However, I don’t think the assessment here is quite right for what local state implies, sorry

No need to apologise for critiquing ideas.  I hold them loosely.  No worries. :)

>  require additional code to strip out [fields that are for local state]

I don't think we disagree here.  That line referred to directive-stripping code.  Stripping the client-only fields needs to happen both with and without directives, so it isn't mentioned in that list.

> I have taken a look at your map and reduce exchanges, but I think they’re unfortunately too small to be published as part of the core. I personally don’t see exchanges as a DSL so introducing another abstraction on top of Wonka streams (which are basically Rx-inspired) would add much value as these exchanges are very small.

Fair assessment.  They add minimal value by themselves.  Both are noop exchanges without wonka operators.  I thought they may be useful for others getting started with exchanges.  It's good to know that you don't want exchanges that tiny in core.

> Basically stateless exchanges are just utilities that don’t have to live in core as they’re very reproducible and just a group of a couple of Wonka operators on top of our convention. IMHO that would add more of a burden to users as there’d be more docs and more concepts that look like conventions they have to adhere to.

Agreed on "don't have to live in core".  I also think that the "very reproducible" part depends on exchange implementation experience.  Speaking as a complete Urql noob trying to write an exchange, I was confused for a while about "simple" things like where request/response happened.  \[1\]

> A local state exchange is also inherently stateful as far as I can tell 🤔

Agreed. And the reduce exchange is stateful. I didn't articulate well that it's also away for me to make sure I have the basic exchange writing parts down before tackling more complex parts. As is, it's like a minimal, bidirectional version of Redux.  It passes state+operation to request and response reducers, and returns the next state+operation.  Combined with wonka operators, it enables any combination of intercepting/merging/creating/filtering requests, responses, and state.  It's a step toward an actual local state implementation.

> I think the way to go here is a full resolver/schema exchange for local state that allows CRUD getters/setters to be defined and strips out fields that it handles before passing these stripped queries on to the fetchExchange / cacheExchange

I can do that.  The getter/setter CRUD approach also differs from what I personally want in a state manager.

How important is having a single local state approach?  One of Urql's biggest strengths IMO is its small core and small composable exchanges.  What about two local state exchanges? One redux-esque implementation and one getter/setter implementation.  Both would strip relevant fields and forward the request.  Given the number of devs experienced with Redux, having both approaches available may also reduce potential barriers to Urql adoption.  Not sure if I'd implement both yet, but I'm happy to start with one and reevaluate afterward.

---------------------------
\[1] _Documentation footnote: Looking back at the exchange creation guide now (thanks for writing that!), I see why I missed where response handling happened.  It mentions "The ExchangeIO function is what each exchange returns. This is a function that receives the stream of operations and must return a stream of results." However, The noopExchange and other examples lack the word "ExchangeIO", as do all the existing exchanges' code I saw.  ExchangeIO seemed like a forgettable part of the docs since it was (seemingly) never used._

_Given that it's quite important and always used anonymously, omitting the exchangeIO type name from the docs would be clearer than including it.  The map exchange, because of its simplicity and explicit request/response semantics, would also be a helpful addition as a "quick start for noobs".  Beyond that, as I mentioned in spectrum, an example that continues after the forward call would also be helpful to clarify how to think about response objects and other post-forward code._

_Once I understand request/response handling accurately, I'd be open to updating that section of the guide from a noob's perspective. Let me know on that._

**Comment by a-laughlin** [comment]
@kitten @JoviDeCroock  Thanks for exploring this with me guys.

My only graphql experience was using the Apollo client, usually with local state.  Frustrations with Apollo's config over composition approach led to exploring Urql.  After seeing the exchanges architecture and small core, I was hooked.  Enough to contribute a local state package.

In actually building that package there are many firsts, like parsing the AST, learning Urql's exchanges and operations, and determining how to modify operations for local state modifications.  While I may have opinions like directive vs directive+resolver approaches, they're based on thinking through the semantic and structural code dependency graphs, not on deep knowledge of graphql query structure (why I missed AST parsing in the original list).  As such, my opinions are likely to have holes specific to graphql and, being human, likely others from time to time. :)

So yes.  I'm very much a noob at this, very open to you pointing out my ill-informed or incorrect thoughts, and very grateful for your insights.  Appreciate you for helping me through this exchange-building process!

**Comment by a-laughlin** [comment]
@kitten I may have incorrectly assumed your "getters/setters" statement implied an OO style.  If your definition includes more functional getters/setters (e.g., lodash's get/set functions), those are fantastic.

My ideal local state API composes those extensively, for example (exact names, functions, args, and structure likely to differ):
```
localStateExchange(
  reduceRequest: schemaLens({
    query:{
      todos:flow( omitFields('x','y'), omitVariables('z') )
    },
    mutation:{
      setTodoXYZ: flow( setState({'x':1}), setVariablesToState('y','z'), preventForward )
    }
  })
)
```

**Comment by a-laughlin** [comment]
The point of this issue is to evaluate what's involved in a local state implementation.  Enumerating specific cases like a good way to wrap it up.  Please let me know if I'm missing any.

If these cases are sufficient, and the above implementation is acceptable enough, we can close this and I'll get started on the tests+code.

**Initialization**
  - set initial local state
  - set request handlers matching schema shape
  - set response handlers matching schema shape

**Request**
  - all
    - map operation to handlers
    - forward operation where operation.query matches no handlers
    - (filter|omit|custom) operation.query properties
    - (get|set|filter|omit|merge|custom) local state properties
    - (get|set|filter|omit|merge|custom) operation[context|variables] properties
    - (get|set|filter|omit|merge|custom) local state properties to operation[context|variables] properties
    - (get|set|filter|omit|merge|custom) operation[context|variables] properties to local state properties
    - request forward (the default)
    - preventForward (like event.preventDefault())
    - generate blank data response (populated in response section)
    - emit blank data response (before forward | after forward | after preventForward)
    - conditionally do any of the above
    - enable async handlers for cases like [localState(sync), cache, localState(async), fetch]
  - query
  - mutation
  - subscription
  - teardown: tear down
  - custom: enable user-defined behaviors

**Response**
  - all
    - map operationResult to handlers
    - return operation where operation.query matches no handlers
    - (get|set|filter|omit|merge|custom) local state properties
    - (get|set|filter|omit|merge|custom) operationResult[data|context] properties
    - (get|set|filter|omit|merge|custom) local state properties to operationResult[data|context] properties
    - (get|set|filter|omit|merge|custom) operationResult[data|context] properties to local state properties
    - map response to [1-n] custom responses
    - preventResponse
    - conditionally do any of the above
    - enable async handlers for cases like [localState(sync), cache, localState(async), fetch]
  - query
  - mutation
  - subscription
  - teardown: tear down
  - custom: enable user-defined behaviors

**Comment by kitten** [comment]
I think I’ll close this for now since we’re not really willing (for now) to maintain more than 2-4 exchanges. It’s actually unlikely that this is going to end up in this repo anyway, but either way, it’d have to be a separate exchange :)

We’re happy to help anyone out though who decides to write one :)

@a-laughlin happy to review any code you’d like to start with! Sorry to close this with your write-up, but I think it’s actually unlikely to be picked up by us and unlikely to be picked up by someone else for the moment.

So if you personally don’t have any capacity I’ll make sure to link to this issue if it comes up again :)

**Comment by a-laughlin** [comment]
Actively working on this.  Description edited to reflect additional learnings in the process, plus some constraints for bundle size and query resolution time.  Let me know if any issues.

**Comment by a-laughlin** [comment]
@kitten Just saw your response.  Thanks for the update.  Issue open or closed are fine with me.

Will you still communicate here if questions come up?

Asking because I do have the capacity for it and it's already in progress.  Current code status is 20-50% done, depending on # of unknown unknowns remaining.  Should have something to post for review in the next few weeks.  Let me know if you'd like me to post it somewhere else.

Thanks for the graphcache reference.  Mid-process learnings from reading that, plus the apollo and apollo and graphql-js implementations have already been useful (and generated a few revisions!).

**Comment by kitten** [comment]
Yea! Happy to communicate here on anything that comes up, or do reach out to me on Twitter if you have very generic/specific questions that aren’t on topic per se. I’m also available on Spectrum (albeit not as promptly) but the alternative is always to just post a thread there that anyone could read 👍

Happy to hear that it’s progressing nicely! ♥️

**Comment by a-laughlin** [comment]
One question.  The graphql spec indicates resolvers should be able to return promises or values.  It makes sense for an `urql` exchange that they should also be able to return streams (i.e., Wonka Sources if my understanding is correct).  Writing predicates for values, promises and functions was straightforward.  Distinguishing a Wonka stream from a regular function is more troublesome.  I couldn't find an example of doing so in the other exchanges or the Wonka source.  How would you write an "isWonkaStream" predicate?

**Comment by kitten** [comment]
@a-laughlin as far as I can tell there isn’t an intended way to differentiate them. You could guarantee this via types and treat functions as always being streams.

That being said, I do not think there is a benefit to using streams there. You’re always returning a single value in those cases, and if you really have a stream you can pick a value off of it using `toPromise` 🤔

**Comment by a-laughlin** [comment]
Yeah.  It's infrequent.  Only one stream-returning case comes to mind: local subscription.  Zero function returning cases come to mind, so your thought about

> treat functions as always being streams 

Should cover it. Thanks!

**Comment by a-laughlin** [comment]
Update and Feedback Request:

An approach using the graphql-js functions looked likely to exceed my targets for bundle size and execution performance. Considering different approaches to solve the problem led to a solution with significant benefits in bundle size, execution speed, learnability, flexibility, and maintainability.  It also has some tradeoffs like a perspective shift and more initial tests.  To me they're worth it.  Curious what you think.

#### Key Concepts
The key perspective shift involves expanding the idea of state to see everything in the operation context as a data tree.  The includes a traditional "state" object, the data object, the operation itself, the query AST, the schema AST, variables, directives, the resolver map... everything.

If everything is a data tree, then every possible state change, from CRUD operations to collection synchronization and omitting fields, are expressible with the same form:

`readSubtrees -> transform -> writeSubtrees`

Composable [lenses](https://medium.com/javascript-scene/lenses-b85976cb0534) can then abstract the tree shape while encapsulating their respective concerns.  As a result, resolvers can use the existing query/mutation api and an identical pipe format for all operations:

`pipe({}, ...readLenses, transform, ...writeLenses)`

The pipe passes a plain object in, readLenses add their respective subtrees' data, transforms  modify desired parts, and writeLenses write the output back to their subtrees.  Transforms just operate on single level collections or objects, so nothing new there.  They can be user defined or come from libraries like lodash.

The last piece is specific to the IO aspect of ExchangeIO.  Resolvers come in pairs.  Here's an example tying everything together:

```js
exchangeFactory({
  data, // imported lenses
  state,
  fields,
  resolverMap:resolverMapLensFactory({
    query:{
      todos:{
        /*  readLenses ->   transform ->                                  writeLenses */
        I:['fields',        omit('local_id'),                              'fields'    ],
        O:['data,state',    flow( addLocalIdTo('data'), unionBy('id')),    'data,state']
      }
    }
  }),
})
```

Let me know if that concept explanation is unclear.  Happy to expand and/or clarify.

#### Benefits
* Load performance: est. < 2KB minified+gzipped for > 98% of use cases.  The core + helpers are currently <300loc.  No external dependencies aside from the wonka peer dependency.  Lenses are separate importable packages, so excluding unneeded lenses (e.g. fields, context, info, type validation) provides granular control over bundle size down to the resolver level.
* Execution perfomance:  est. < 3ms query/mutation resolution for local-only data.  Likely better. 
 Lens exclusion also directly improves execution speed.
* Learnability:  Assuming prior understanding of trees, queries/mutations, and the exchange pipeline, there's almost nothing to learn with this this approach.  The existing query/mutation api plus resolver pairs are sufficient for all operations.  Transforms leverage existing user knowledge.  Learning lenses for most cases is equivalent to learning "the data lens reads and writes the data property".
* Flexibility: Lenses can import and compose other lenses.  Resolvers' transform functions may return values, promises, or streams.  Lenses and transforms take the same single plain object argument, making all non-core functionality composable.
* Maintainability:  The core is also a pipe of 5 single-responsibility functions totaling < 100loc. There is no conditional logic in the core or lenses, one argument per function, and no functions longer than 30 lines.  New features are new lenses, not core changes, so maintenance on existing code should be quite low once lenses for main use cases are written.


#### Tradeoffs 
* Seeing everything as a tree requires a mindset shift.
* It's possible that lenses can't handle some case, but I haven't been to think of one yet, especially since they all take the same argument, and can import and compose others lenses' behaviors.
* Resolver pairs differ from other resolver implementations.  I personally find them more intuitive to work with for bidirectional operations, but others may not.  Swapping that syntax involves swapping a 10 function in the core pipe.
* Some folks will likely have a negative reaction to the string syntax for lenses.  I specifically made them that way because they represent paths in the tree, they intentionally lack an api, and memoization is simpler and faster with strings.  If folks really care, swapping out that core function is easy too.
* Traditional resolvers will need an adapter to work.  Like queries, the exchange provides lens data declared in the resolver.  Context, info, and variables aren't imported by default.  An adapter to import those and spread them to the `(parent, variables, context, info)` format is trivial, so there's an option to use existing resolvers.  Bidirectional operations will still need a pair.
* The biggest tradeoff for me is that enabling resolvers to work on the expanded definition of state means implementing a decent chunk of the spec from scratch.  That's surprisingly not too bad with type validation and nullability as separate composable concerns.  They may even be unnecessary, or addable later, if other exchanges or server implementations also do that.  Either way the inability to to rely on graphql-js means more tests up front.  Still, this architecture's benefits outweigh that cost for me.

**Comment by a-laughlin** [comment]
Work and family take priority for the next few weeks.  Otherwise continuing forward with this architecture.  Will post a repo once core lenses and tests are in place.

Hope you all have fantastic holidays!

**Comment by anilanar** [comment]
@a-laughlin Were you able to make any progress?

**Comment by a-laughlin** [comment]
Hi Anil,

It's still in progress, though at a slow pace due to higher priority work issues.  The biggest challenge thus far was getting typescript to work with its higher order functions.  Not sure a  typescript version will ultimately be worth the effort.  My plan at this point is to re-look at it next week and decide whether to keep pursuing the TS version or get it done in JS.
